Dockerfile

```diff
@@ -0,0 +1,23 @@
+FROM python:3.9-slim-bullseye
+
+# Copy minimal set of SQLFluff install files
+# into their own folder for easier debugging.
+WORKDIR /app
+COPY src ./src
+COPY setup.py .
+COPY MANIFEST.in .
+COPY README.md .
+
+# Install SQLFluff in virtual environment to
+# avoid potential clashes with system dependencies.
+RUN python -m venv .venv \
+    && . .venv/bin/activate \
+    && pip install --upgrade pip setuptools wheel \
+    && pip install .
+
+# Switch to non-root user.
+USER 5000
+
+# Set SQLFluff command as entry point for image.
+ENTRYPOINT ["/app/.venv/bin/sqlfluff"]
+CMD ["--help"]
```

Comment:
```suggestion
ENV VIRTUAL_ENV .venv
RUN python -m venv $VIRTUAL_ENV
ENV PATH $VIRTUAL_ENV/bin:$PATH
RUN pip install --upgrade pip setuptools wheel
RUN pip install .

# Switch to non-root user.
USER 5000

# Set SQLFluff command as entry point for image.
ENTRYPOINT ["sqlfluff"]
CMD ["--help"]
```

By activating the virtual environment through `ENV` statements, you should be able to avoid the fully qualified pathname in `ENTRYPOINT` and can split up the `pip` commands over multiple `RUN` statements. 

Further, you could push the virtual environment creation up towards the top of the Dockerfile (above the `COPY` statements), along with `pip install --upgrade` command which would better take advantage of docker layer caching. Were someone doing incremental development with the docker images, this would save some build time. Though, I admit, I don't think it would be terribly substantial time savings.

Feel free to keep the way as is, if intended/preferred of course.

Reply:
Following up on this further - I think there's potential for substantial build time savings. You should be able to cache the dependencies between builds if none of the application code in `src` has changed. 

It would look something like this:

```Dockerfile
WORKDIR /app

ENV VIRTUAL_ENV .venv
RUN python -m venv $VIRTUAL_ENV
ENV PATH $VIRTUAL_ENV/bin:$PATH
RUN pip install --upgrade pip setuptools wheel

COPY setup.py .

RUN python setup.py egg_info
RUN pip install -r *.egg-info/requires.txt

COPY src ./src
COPY MANIFEST.in .
COPY README.md .

# continue with rest of file
```

The gotcha here is that it looks like `setup.py` depends on `src`...so it may take extra effort to realize this correctly. And maybe that makes it not worth the time savings :shrug: 

I'll note that GitHub Actions supports docker layer caching, so it can save time there as well if configured correctly. Maybe that would (over time) have a meaningful impact on total free minutes used on GitHub Actions. I'm new to the project so I don't know how concerned the group is about that sort of stuff.

See: https://github.com/docker/build-push-action/blob/master/docs/advanced/cache.md

Reply:
See https://github.com/jpy-git/sqlfluff/pull/1 for a functional version of `Dockerfile` akin to what I'm suggesting

Reply:
@ccurro amazing thanks for the input! 😄 Will have a play around with the alternate venv activation, and agree that it makes sense to have the initial venv setup before copying over package.

Reply:
On the second point, for context this image will be built on release of a new sqlfluff version, which is roughly every 1-2 weeks.
These new releases will always be for enhancements/new features/bugfixes, all of which mean changes to code in the src folder so unfortunately the method shown above for caching dependencies won't work.
However, we do have a requirements.txt file that I can use in the dependencies step to include them in cached layers. I think the ideal solution for this is to move from setup.py to setup.cfg in the future as we can then easily extract the dependency metadata from there as a single source of truth.
I'll take a look at the GHA caching mentioned as well. When I was testing the GHA build was only taking ~4 mins and as I said before will only be run roughly weekly so not a huge need to increase speed/complexity.

Reply:
@ccurro I've made the changes following the discussion above:
- Virtual environment is created at the start and prepended to the path.
- Install requirements separately to take advantage of improved layer caching. (Currently from requirements.txt but I am going to look at adding setup.cfg to extract requirements from)
- Incorporate inline caching into GitHub action workflow. first build is still 4 mins (since there's no cache), but then subsequent builds (with no change to requirements) take 1 min.

Think the image is looking very nice now, thanks 😄

Reply:
Looks great to me! :+1: